### PR TITLE
feat: 11231: Improved virtual node remover

### PR DESCRIPTION
**Pull request by artemananiev**
Fix summary:

* Re-implemented `VirtualNodeRemover` without virtual key sets, bloom filters, etc.
* Added support for `putIfEquals()`/`deleteIfEquals()` to `HalfDiskHashMap` and related classes. This functionality is used by the new node remover
* Removed all classes no longer used by node remover, e.g. `VirtualKeySet`
* Removed all methods no longer used by node remover, e.g. `VirtualDataSource.buildKeySet()`
* A couple new unit tests are added to cover node remover corner cases

Fixes: https://github.com/hashgraph/hedera-services/issues/11231
Signed-off-by: Artem Ananev <artem.ananev@swirldslabs.com>

### CODE REVIEW

File: platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java [hunk 1]
```diff
@@ -621,6 +599,8 @@ public VirtualLeafRecord<K, V> loadLeafRecord(final K key) throws IOException {
             return null;
         }
 
+        //        assert validLeafPathRange.withinRange(path);
```
Comment (imalygin) [inline platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java:602]: Please remove it if it's no longer relevant

File: platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java [hunk 2]
```diff
@@ -1230,11 +1210,15 @@ private void writeLeavesToPathToKeyValue(
 
         // iterate over leaf records to delete
         deletedLeaves.forEach(leafRecord -> {
+            final long path = leafRecord.getPath();
             // update objectKeyToPath
             if (isLongKeyMode) {
-                longKeyToPath.put(((VirtualLongKey) leafRecord.getKey()).getKeyAsLong(), INVALID_PATH);
+                final long key = ((VirtualLongKey) leafRecord.getKey()).getKeyAsLong();
+                //                longKeyToPath.put(((VirtualLongKey) leafRecord.getKey()).getKeyAsLong(),
```
Comment (imalygin) [inline platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java:1217]: Please remove

File: platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java [hunk 3]
```diff
@@ -1230,11 +1210,15 @@ private void writeLeavesToPathToKeyValue(
...
+                // INVALID_PATH);
+                longKeyToPath.putIfEqual(key, path, INVALID_PATH);
             } else {
-                objectKeyToPath.delete(leafRecord.getKey());
+                objectKeyToPath.deleteIfEquals(leafRecord.getKey(), path);
```
Comment (imalygin) [inline platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java:1221]: nit: `putIfEqual` but `deleteIfEquals`. Could you please make sure that the naming is consistent?

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/reconnect/ReconnectNodeRemover.java [hunk 4]
```diff
@@ -195,187 +119,50 @@ public void setPathInformation(final long newFirstLeafPath, final long newLastLe
...
-        final Stream<VirtualLeafRecord<K, V>> stream = keysToBeRemoved.entrySet().stream()
-                .map((final Map.Entry<K, Long> entry) ->
-                        new VirtualLeafRecord<>(entry.getValue(), entry.getKey(), null));
-
-        // We can't just clear the map, as doing so will disrupt the stream constructed above.
-        keysToBeRemoved = new HashMap<>();
-
-        workQueue.resume();
-
+    public Stream<VirtualLeafRecord<K, V>> getRecordsToDelete() {
```
Comment (imalygin) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/reconnect/ReconnectNodeRemover.java:161]: Usually getters are idempotent. Please rename it or, at least, mention in the javadoc that it's not idempotent.

File: platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java (hunk 1 above)
Comment (artemananiev) [reply platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java:602]: Removed.

File: platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java (hunk 2 above)
Comment (artemananiev) [reply platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java]: Removed

File: platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java (hunk 3 above)
Comment (artemananiev) [reply platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java:1221]: This `putIfEqual()` existed before this change, it's a part of `CASableLongIndex` interface. It makes sense to introduce the new method with the same name for consistency. I'll change everything to `putIfEqual()` and `deleteIfEqual()`

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/reconnect/ReconnectNodeRemover.java (hunk 4 above)
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/reconnect/ReconnectNodeRemover.java:161]: I added a comment to this method and also to `leavesToDelete` field in the same class.

File: platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/hashmap/BucketMutation.java [hunk 5]
```diff
@@ -59,6 +59,23 @@ void put(K key, long value) {
...
     }
 
+    void putIfEquals(final K key, final long oldValue, final long value) {
+        BucketMutation<K> mutation = this;
+        while (true) {
+            if (mutation.key.equals(key)) {
+                if (mutation.value == oldValue) {
+                    mutation.value = value;
+                }
+                return;
```
[the remark this answers is not in the document]
Comment (artemananiev) [reply platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/hashmap/BucketMutation.java:83]: No more than one mutation is expected for a key. If there are multiple `put()`s for the same key, they will result in a single mutation (with the value from the last `put()` request). This is why I return from this method regardless of old value to be equal or not to the current value, as soon as a mutation for the key is encountered.

File: platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/hashmap/HalfDiskHashMap.java [hunk 6]
```diff
@@ -333,11 +326,26 @@ public void put(final K key, final long value) {
...
+     *
+     * @param key the key to store the value for
+     * @param value the value to store for given key
+     */
+    public void put(final K key, final long value) {
+        final BucketMutation<K> bucketMap = findBucketForUpdate(key, value);
         bucketMap.put(key, value);
     }
 
+    public void putIfEquals(final K key, final long oldValue, final long value) {
```
[the remark this answers is not in the document]
Comment (artemananiev) [reply platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/hashmap/HalfDiskHashMap.java:344]: Added JavaDoc for both `putIfEqual()` and `deleteIfEqual()`

File: platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/hashmap/HalfDiskHashMap.java [hunk 7]
```diff
@@ -341,22 +348,49 @@ public void put(final K key, final long value) {
         bucketMap.put(key, value);
     }
 
-    public void putIfEquals(final K key, final long oldValue, final long value) {
+    /**
+     * Put a key/valur during the current writing session. This method is similar to {@link
```
Comment (jsync-swirlds) [inline platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/hashmap/HalfDiskHashMap.java:352]:
```suggestion
     * Put a key/value during the current writing session. This method is similar to {@link
```

Comment (artemananiev) [reply platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/hashmap/HalfDiskHashMap.java:352]: Good catch. Fixed

**Review by imalygin — APPROVED** [review]

**Review by anthony-swirldslabs — APPROVED** [review]

**Review by jsync-swirlds — APPROVED** [review]

**Review by OlegMazurov — APPROVED** [review]
Tested reconnect in single-node mode. NFT transfer benchmark (10M accounts/100M NFTs) was running at ~7-8K TPS. Full reconnect took 1760 sec, subsequent partial reconnect - 1144 sec.
